File: Quick Starts/Cryptography/VB-UnityIntegration/CryptographyQuickStart/QuickStartForm.vb
'===============================================================================
' Microsoft patterns & practices Enterprise Library
' Cryptography Application Block QuickStart
'===============================================================================
' Copyright ? Microsoft Corporation.  All rights reserved.
' THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY
' OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT
' LIMITED TO THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
' FITNESS FOR A PARTICULAR PURPOSE.
'===============================================================================


Public Class QuickStartForm
    Inherits System.Windows.Forms.Form

    Private viewerProcess As Process = Nothing
    Private Const HelpViewerArguments As String = "/helpcol ms-help://MS.VSCC.v90/MS.VSIPCC.v90/ms.practices.entlib.2008may /LaunchFKeywordTopic CryptographyQS1"

    ' The following strings correspond to the names of the providers as 
    ' specified by the configuration file.
    Private Const hashProvider As String = "hashprovider"
	Private Const symmProvider As String = "symprovider"

	Private Const symmKeyFileName As String = "SymmetricKeyFile.txt"

    ' Encrypted string
    Private stringToEncrypt As String
    Private encryptedContentsBase64 As String

    ' Generated hash
    Private stringForHash As String
    Private generatedHash As Byte()

    Private cryptographyManager As CryptographyManager

    Private inputFormObject As InputForm = New InputForm

    Public Shared AppForm As System.Windows.Forms.Form

#Region " Windows Form Designer generated code "

    Public Sub New()
        MyBase.New()

        'This call is required by the Windows Form Designer.
        InitializeComponent()
    End Sub

    'Form overrides dispose to clean up the component list.
    Protected Overloads Overrides Sub Dispose(ByVal disposing As Boolean)
        If disposing Then
            If Not (components Is Nothing) Then
                components.Dispose()
            End If
        End If
        MyBase.Dispose(disposing)
    End Sub

    'Required by the Windows Form Designer
    Private components As System.ComponentModel.IContainer

    'NOTE: The following procedure is required by the Windows Form Designer
    'It can be modified using the Windows Form Designer.  
    'Do not modify it using the code editor.
    Friend WithEvents checkTextButton As System.Windows.Forms.Button
    Friend WithEvents getHashButton As System.Windows.Forms.Button
    Friend WithEvents decryptButton As System.Windows.Forms.Button
    Friend WithEvents groupBox As System.Windows.Forms.GroupBox
    Friend WithEvents viewWalkthroughButton As System.Windows.Forms.Button
    Friend WithEvents quitButton As System.Windows.Forms.Button
    Friend WithEvents groupBox1 As System.Windows.Forms.GroupBox
    Friend WithEvents label2 As System.Windows.Forms.Label
    Friend WithEvents logoPictureBox As System.Windows.Forms.PictureBox
    Friend WithEvents resultsTextBox As System.Windows.Forms.TextBox
    Friend WithEvents encryptButton As System.Windows.Forms.Button
    <System.Diagnostics.DebuggerStepThrough()> Private Sub InitializeComponent()
        Dim resources As System.ComponentModel.ComponentResourceManager = New System.ComponentModel.ComponentResourceManager(GetType(QuickStartForm))
        Me.checkTextButton = New System.Windows.Forms.Button
        Me.getHashButton = New System.Windows.Forms.Button
        Me.decryptButton = New System.Windows.Forms.Button
        Me.groupBox = New System.Windows.Forms.GroupBox
        Me.viewWalkthroughButton = New System.Windows.Forms.Button
        Me.quitButton = New System.Windows.Forms.Button
        Me.groupBox1 = New System.Windows.Forms.GroupBox
        Me.label2 = New System.Windows.Forms.Label
        Me.logoPictureBox = New System.Windows.Forms.PictureBox
        Me.resultsTextBox = New System.Windows.Forms.TextBox
        Me.encryptButton = New System.Windows.Forms.Button
        Me.groupBox.SuspendLayout()
        Me.groupBox1.SuspendLayout()
        CType(Me.logoPictureBox, System.ComponentModel.ISupportInitialize).BeginInit()
        Me.SuspendLayout()
        '
        'checkTextButton
        '
        resources.ApplyResources(Me.checkTextButton, "checkTextButton")
        Me.checkTextButton.Name = "checkTextButton"
        '
        'getHashButton
        '
        resources.ApplyResources(Me.getHashButton, "getHashButton")
        Me.getHashButton.Name = "getHashButton"
        '
        'decryptButton
        '
        resources.ApplyResources(Me.decryptButton, "decryptButton")
        Me.decryptButton.Name = "decryptButton"
        '
        'groupBox
        '
        Me.groupBox.Controls.Add(Me.viewWalkthroughButton)
        Me.groupBox.Controls.Add(Me.quitButton)
        resources.ApplyResources(Me.groupBox, "groupBox")
        Me.groupBox.Name = "groupBox"
        Me.groupBox.TabStop = False
        '
        'viewWalkthroughButton
        '
        resources.ApplyResources(Me.viewWalkthroughButton, "viewWalkthroughButton")
        Me.viewWalkthroughButton.Name = "viewWalkthroughButton"
        '
        'quitButton
        '
        resources.ApplyResources(Me.quitButton, "quitButton")
        Me.quitButton.Name = "quitButton"
        '
        'groupBox1
        '
        Me.groupBox1.BackColor = System.Drawing.Color.White
        Me.groupBox1.Controls.Add(Me.label2)
        Me.groupBox1.Controls.Add(Me.logoPictureBox)
        resources.ApplyResources(Me.groupBox1, "groupBox1")
        Me.groupBox1.Name = "groupBox1"
        Me.groupBox1.TabStop = False
        '
        'label2
        '
        resources.ApplyResources(Me.label2, "label2")
        Me.label2.Name = "label2"
        '
        'logoPictureBox
        '
        resources.ApplyResources(Me.logoPictureBox, "logoPictureBox")
        Me.logoPictureBox.Name = "logoPictureBox"
        Me.logoPictureBox.TabStop = False
        '
        'resultsTextBox
        '
        resources.ApplyResources(Me.resultsTextBox, "resultsTextBox")
        Me.resultsTextBox.Name = "resultsTextBox"
        Me.resultsTextBox.ReadOnly = True
        Me.resultsTextBox.TabStop = False
        '
        'encryptButton
        '
        resources.ApplyResources(Me.encryptButton, "encryptButton")
        Me.encryptButton.Name = "encryptButton"
        '
        'QuickStartForm
        '
        resources.ApplyResources(Me, "$this")
        Me.Controls.Add(Me.checkTextButton)
        Me.Controls.Add(Me.getHashButton)
        Me.Controls.Add(Me.decryptButton)
        Me.Controls.Add(Me.groupBox)
        Me.Controls.Add(Me.groupBox1)
        Me.Controls.Add(Me.resultsTextBox)
        Me.Controls.Add(Me.encryptButton)
        Me.FormBorderStyle = System.Windows.Forms.FormBorderStyle.FixedSingle
        Me.MaximizeBox = False
        Me.Name = "QuickStartForm"
        Me.groupBox.ResumeLayout(False)
        Me.groupBox1.ResumeLayout(False)
        Me.groupBox1.PerformLayout()
        CType(Me.logoPictureBox, System.ComponentModel.ISupportInitialize).EndInit()
        Me.ResumeLayout(False)
        Me.PerformLayout()

    End Sub

#End Region

    Public Sub New(ByVal cryptographyManager As CryptographyManager)
        Me.New()
        Me.cryptographyManager = cryptographyManager
    End Sub

    Public Shared Sub CreateKeys()
        Dim installedPath As String = ConfigurationManager.AppSettings("InstallPath")
        Dim fileName As String = Path.Combine(installedPath, symmKeyFileName)
        Dim symmetricKey As ProtectedKey = KeyManager.GenerateSymmetricKey(GetType(RijndaelManaged), DataProtectionScope.LocalMachine)
        Using keyStream As FileStream = New FileStream(fileName, FileMode.Create)
            KeyManager.Write(keyStream, symmetricKey)
        End Using
    End Sub

    ' Displays dialog with information about exceptions that occur in the application. 
    Private Shared Sub AppThreadException(ByVal source As Object, ByVal e As ThreadExceptionEventArgs)
        Dim errorMsg As String = String.Format(New CultureInfo("en-us", True), "There are some problems while trying to use the Cryptography Quick Start, please check the following error messages: " & Environment.NewLine & "{0}" & Environment.NewLine, e.Exception.Message)
        errorMsg = errorMsg & Environment.NewLine

        Dim result As DialogResult = MessageBox.Show(errorMsg, "Application Error", MessageBoxButtons.AbortRetryIgnore, MessageBoxIcon.Stop)

        ' Exits the program when the user clicks Abort.
        If (result = System.Windows.Forms.DialogResult.Abort) Then
            Application.Exit()
        End If

        QuickStartForm.AppForm.Cursor = Cursors.Default
    End Sub

    Private Function GetEmbeddedImage(ByVal resourceName As String) As Image
        Dim resourceStream As Stream = System.Reflection.Assembly.GetEntryAssembly().GetManifestResourceStream(resourceName)

        If (resourceStream Is Nothing) Then
            Return Nothing
        End If

        Dim img As Image = Image.FromStream(resourceStream)

        Return img
    End Function

    Private Sub DisplayResults(ByVal results As String)
        Me.resultsTextBox.Text &= results & Environment.NewLine
        Me.resultsTextBox.SelectAll()
        Me.resultsTextBox.ScrollToCaret()
    End Sub

    Private Sub QuickStartForm_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        ' Initialize image to embedded logo
        Me.logoPictureBox.Image = GetEmbeddedImage("logo.gif")

        Me.decryptButton.Enabled = False
        Me.checkTextButton.Enabled = False
    End Sub

    Private Sub quitButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles quitButton.Click
        Me.Close()
    End Sub

    ' Encrypt a sample text using the default symmetric provider.
    Private Sub encryptButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles encryptButton.Click
        If (Not inputFormObject Is Nothing) Then
            inputFormObject.Title = My.Resources.EncryptTitleMessage
            inputFormObject.InstructionsText = My.Resources.EncryptInstructionsMessage
            If (inputFormObject.ShowDialog() = Windows.Forms.DialogResult.OK) Then
                Try
                    Me.Cursor = Cursors.WaitCursor

                    Me.encryptedContentsBase64 = cryptographyManager.EncryptSymmetric(symmProvider, inputFormObject.Input)

                    Me.DisplayResults(String.Format(My.Resources.Culture, My.Resources.OriginalTextMessage, inputFormObject.Input))
                    Me.DisplayResults(String.Format(My.Resources.EncryptedTextMessage, Me.encryptedContentsBase64))

                    Me.decryptButton.Enabled = True

                    Me.Cursor = Cursors.Arrow
                Catch ex As Exception
                    ProcessUnhandledException(ex)
                Finally
                    Me.Cursor = Cursors.Default
                End Try
            End If
        End If
    End Sub

    ' Decrypts a set of bytes and displayed the decrypted contents.
    Private Sub decryptButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles decryptButton.Click
        If (Not Me.encryptedContentsBase64 Is Nothing) Then

            Try
                Me.Cursor = Cursors.WaitCursor

                ' Get a string for display
                Dim readableString As String = cryptographyManager.DecryptSymmetric(symmProvider, Me.encryptedContentsBase64)

                Me.DisplayResults(String.Format(My.Resources.Culture, My.Resources.DecryptedTextMessage, readableString))
            Catch ex As Exception
                ProcessUnhandledException(ex)
            Finally
                Me.Cursor = Cursors.Default
            End Try

        Else
            Me.DisplayResults(My.Resources.DecryptErrorMessage)
        End If
    End Sub

    ' Creates a hash based on a sample text, for further comparison.
    Private Sub getHashButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles getHashButton.Click
        If (Not inputFormObject Is Nothing) Then

            inputFormObject.Title = My.Resources.HashTitleMessage
            inputFormObject.InstructionsText = My.Resources.HashInstructionsMessage
            If (inputFormObject.ShowDialog() = Windows.Forms.DialogResult.OK) Then

                Try
                    Me.Cursor = Cursors.WaitCursor

                    Me.generatedHash = Me.GetHash(inputFormObject.Input)

                    Me.DisplayResults(String.Format(My.Resources.Culture, My.Resources.HashMessage, Convert.ToBase64String(generatedHash)))

                    Me.checkTextButton.Enabled = True
                Catch ex As Exception
                    ProcessUnhandledException(ex)
                Finally
                    Me.Cursor = Cursors.Default
                End Try
            End If
        End If
    End Sub

    Private Function GetHash(ByVal plainText As String) As Byte()

        Dim valueToHash As Byte() = Encoding.UTF8.GetBytes(plainText)

        Dim generatedHash As Byte() = cryptographyManager.CreateHash(hashProvider, valueToHash)

        ' Clear the byte array memory
        Array.Clear(valueToHash, 0, valueToHash.Length)

        Return generatedHash

    End Function

    Private Function CompareHash(ByVal plainText As String, ByVal existingHashValue As Byte()) As Boolean

        Dim valueToHash As Byte() = Encoding.UTF8.GetBytes(plainText)

        Dim matched As Boolean = cryptographyManager.CompareHash(hashProvider, valueToHash, existingHashValue)

        ' Clear the byte array memory
        Array.Clear(valueToHash, 0, valueToHash.Length)

        Return matched
    End Function

    ' Check a previously generated hash in order to determine whether the original text
    ' has been tampered or not.
    Private Sub checkTextButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles checkTextButton.Click
        If (Not inputFormObject Is Nothing) Then
            inputFormObject.Title = My.Resources.HashTitleMessage
            inputFormObject.InstructionsText = My.Resources.HashInstructionsMessage
            If (inputFormObject.ShowDialog() = Windows.Forms.DialogResult.OK) Then
                If (Not Me.generatedHash Is Nothing) Then
                    Try
                        Me.Cursor = Cursors.WaitCursor

                        Dim comparisonSucceeded As Boolean = Me.CompareHash(inputFormObject.Input, Me.generatedHash)

                        If (comparisonSucceeded) Then
                            Me.DisplayResults(My.Resources.TextNotTamperedMessage)
                        Else
                            Me.DisplayResults(My.Resources.TextTamperedMessage)
                        End If

                    Catch ex As Exception
                        ProcessUnhandledException(ex)
                    Finally
                        Me.Cursor = Cursors.Default
                    End Try
                End If
            End If
        End If
    End Sub
    Private Function GetHelpViewerExecutable() As String
        Dim common As String = Environment.GetEnvironmentVariable("CommonProgramFiles")
        Return Path.Combine(common, "Microsoft Shared\Help 9\dexplore.exe")
    End Function
    Private Sub viewWalkthroughButton_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles viewWalkthroughButton.Click
        ' Process has never been started. Initialize and launch the viewer.
        If (Me.viewerProcess Is Nothing) Then
            ' Initialize the Process information for the help viewer
            Me.viewerProcess = New Process

            Me.viewerProcess.StartInfo.FileName = GetHelpViewerExecutable()
            Me.viewerProcess.StartInfo.Arguments = HelpViewerArguments
            Me.viewerProcess.Start()
        ElseIf (Me.viewerProcess.HasExited) Then
            ' Process previously started, then exited. Start the process again.
            Me.viewerProcess.Start()
        Else
            ' Process was already started - bring it to the foreground
            Dim hWnd As IntPtr = Me.viewerProcess.MainWindowHandle
            If (NativeMethods.IsIconic(hWnd)) Then
                NativeMethods.ShowWindowAsync(hWnd, NativeMethods.SW_RESTORE)
            End If
            NativeMethods.SetForegroundWindow(hWnd)
        End If

    End Sub

    ''' <summary>
    ''' Process any unhandled exceptions that occur in the application.
    ''' This code is called by all UI entry points in the application (e.g. button click events)
    ''' when an unhandled exception occurs.
    ''' You could also achieve this by handling the Application.ThreadException event, however
    ''' the VS2005 debugger will break before this event is called.
    ''' </summary>
    ''' <param name="ex">The unhandled exception</param>
    Private Sub ProcessUnhandledException(ByVal ex As Exception)
        Dim errorMessage As StringBuilder = New StringBuilder()
        errorMessage.Append("The following error occured during execution of the Cryptography QuickStart.")
        errorMessage.Append(Environment.NewLine & Environment.NewLine)
        errorMessage.AppendFormat(New CultureInfo("en-us", True), "{0}", ex.Message)
        errorMessage.Append(Environment.NewLine & Environment.NewLine)
        errorMessage.Append("Exceptions can be caused by invalid configuration information.")
        errorMessage.Append(Environment.NewLine & Environment.NewLine)
        errorMessage.Append("Do you want to exit the application?")

        Dim result As Windows.Forms.DialogResult = MessageBox.Show(errorMessage.ToString(), "Application Error", MessageBoxButtons.YesNo, MessageBoxIcon.Stop)

        ' Exits the program when the user clicks Abort.
        If (result = Windows.Forms.DialogResult.Yes) Then
            Application.Exit()
        End If
    End Sub
End Class

File: Quick Starts/ExceptionHandling/CS/ExceptionHandlingWithLoggingQuickStart/QuickStartForm.cs
//===============================================================================
// Microsoft patterns & practices Enterprise Library
// Exception Handling Application Block QuickStart
//===============================================================================
// Copyright © Microsoft Corporation.  All rights reserved.
// THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY
// OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT
// LIMITED TO THE IMPLIED WARRANTIES OF MERCHANTABILITY AND
// FITNESS FOR A PARTICULAR PURPOSE.
//===============================================================================

using System;
using System.Diagnostics;
using System.IO;
using System.Reflection;
using System.Text;
using System.Windows.Forms;

using ExceptionHandlingQuickStart.BusinessLayer;
using Microsoft.Practices.EnterpriseLibrary.ExceptionHandling;

namespace ExceptionHandlingQuickStart
{
    /// <summary>
    /// Enterprise Library Exception Handling Application Block Quick Start.
    /// </summary>
    public class QuickStartForm : System.Windows.Forms.Form
    {
        /// <summary>
        /// Required designer variable.
        /// </summary>
        private System.ComponentModel.Container components = null;
        private System.Windows.Forms.GroupBox groupBox1;
        private System.Windows.Forms.Label label2;
        private System.Windows.Forms.GroupBox groupBox;

        private Process viewerProcess = null;
        private System.Windows.Forms.Button logExceptionButton;
        private System.Windows.Forms.Button notifyUserButton;
        private System.Windows.Forms.Button viewWalkthroughButton;
        private System.Windows.Forms.Button quitButton;
        private System.Windows.Forms.PictureBox logoPictureBox;
        private System.Windows.Forms.TextBox resultsTextBox;

        private const string HelpViewerArguments = @"/helpcol ms-help://MS.VSCC.v90/MS.VSIPCC.v90/ms.practices.entlib.2008oct /LaunchFKeywordTopic ExceptionhandlingQS2";

        public static System.Windows.Forms.Form AppForm;

        public QuickStartForm()
        {
            //
            // Required for Windows Form Designer support
            //
            InitializeComponent();
        }

        /// <summary>
        /// Clean up any resources being used.
        /// </summary>
        protected override void Dispose(bool disposing)
        {
            if (disposing)
            {
                if (components != null)
                {
                    components.Dispose();
                }
            }
            base.Dispose(disposing);
        }

        #region Windows Form Designer generated code
        /// <summary>
        /// Required method for Designer support - do not modify
        /// the contents of this method with the code editor.
        /// </summary>
        private void InitializeComponent()
        {
            this.resultsTextBox = new System.Windows.Forms.TextBox();
            this.logExceptionButton = new System.Windows.Forms.Button();
            this.groupBox1 = new System.Windows.Forms.GroupBox();
            this.label2 = new System.Windows.Forms.Label();
            this.logoPictureBox = new System.Windows.Forms.PictureBox();
            this.groupBox = new System.Windows.Forms.GroupBox();
            this.viewWalkthroughButton = new System.Windows.Forms.Button();
            this.quitButton = new System.Windows.Forms.Button();
            this.notifyUserButton = new System.Windows.Forms.Button();
            this.groupBox1.SuspendLayout();
            ((System.ComponentModel.ISupportInitialize)(this.logoPictureBox)).BeginInit();
            this.groupBox.SuspendLayout();
            this.SuspendLayout();
            // 
            // resultsTextBox
            // 
            this.resultsTextBox.Location = new System.Drawing.Point(192, 88);
            this.resultsTextBox.Multiline = true;
            this.resultsTextBox.Name = "resultsTextBox";
            this.resultsTextBox.ReadOnly = true;
            this.resultsTextBox.ScrollBars = System.Windows.Forms.ScrollBars.Vertical;
            this.resultsTextBox.Size = new System.Drawing.Size(614, 272);
            this.resultsTextBox.TabIndex = 3;
            this.resultsTextBox.TabStop = false;
            // 
            // logExceptionButton
            // 
            this.logExceptionButton.FlatStyle = System.Windows.Forms.FlatStyle.System;
            this.logExceptionButton.Location = new System.Drawing.Point(19, 96);
            this.logExceptionButton.Name = "logExceptionButton";
            this.logExceptionButton.Size = new System.Drawing.Size(154, 46);
            this.logExceptionButton.TabIndex = 1;
            this.logExceptionButton.Text = "&Log an exception";
            this.logExceptionButton.Click += new System.EventHandler(this.logExceptionButton_Click);
            // 
            // groupBox1
            // 
            this.groupBox1.BackColor = System.Drawing.Color.White;
            this.groupBox1.Controls.Add(this.label2);
            this.groupBox1.Controls.Add(this.logoPictureBox);
            this.groupBox1.FlatStyle = System.Windows.Forms.FlatStyle.System;
            this.groupBox1.Location = new System.Drawing.Point(0, -9);
            this.groupBox1.Name = "groupBox1";
            this.groupBox1.Size = new System.Drawing.Size(845, 83);
            this.groupBox1.TabIndex = 0;
            this.groupBox1.TabStop = false;
            // 
            // label2
            // 
            this.label2.Font = new System.Drawing.Font("Microsoft Sans Serif", 15.75F, System.Drawing.FontStyle.Regular, System.Drawing.GraphicsUnit.Point, ((byte)(0)));
            this.label2.Location = new System.Drawing.Point(19, 28);
            this.label2.Name = "label2";
            this.label2.Size = new System.Drawing.Size(517, 35);
            this.label2.TabIndex = 1;
            this.label2.Text = "Exception Handling With Logging";
            // 
            // logoPictureBox
            // 
            this.logoPictureBox.Location = new System.Drawing.Point(730, 16);
            this.logoPictureBox.Name = "logoPictureBox";
            this.logoPictureBox.Size = new System.Drawing.Size(69, 50);
            this.logoPictureBox.SizeMode = System.Windows.Forms.PictureBoxSizeMode.AutoSize;
            this.logoPictureBox.TabIndex = 0;
            this.logoPictureBox.TabStop = false;
            // 
            // groupBox
            // 
            this.groupBox.Controls.Add(this.viewWalkthroughButton);
            this.groupBox.Controls.Add(this.quitButton);
            this.groupBox.FlatStyle = System.Windows.Forms.FlatStyle.System;
            this.groupBox.Location = new System.Drawing.Point(0, 376);
            this.groupBox.Name = "groupBox";
            this.groupBox.Size = new System.Drawing.Size(845, 101);
            this.groupBox.TabIndex = 4;
            this.groupBox.TabStop = false;
            // 
            // viewWalkthroughButton
            // 
            this.viewWalkthroughButton.FlatStyle = System.Windows.Forms.FlatStyle.System;
            this.viewWalkthroughButton.Location = new System.Drawing.Point(528, 28);
            this.viewWalkthroughButton.Name = "viewWalkthroughButton";
            this.viewWalkthroughButton.Size = new System.Drawing.Size(125, 37);
            this.viewWalkthroughButton.TabIndex = 0;
            this.viewWalkthroughButton.Text = "View &Walkthrough";
            this.viewWalkthroughButton.Click += new System.EventHandler(this.viewWalkthroughButton_Click);
            // 
            // quitButton
            // 
            this.quitButton.DialogResult = System.Windows.Forms.DialogResult.Cancel;
            this.quitButton.FlatStyle = System.Windows.Forms.FlatStyle.System;
            this.quitButton.Location = new System.Drawing.Point(672, 28);
            this.quitButton.Name = "quitButton";
            this.quitButton.Size = new System.Drawing.Size(125, 37);
            this.quitButton.TabIndex = 1;
            this.quitButton.Text = "&Quit";
            this.quitButton.Click += new System.EventHandler(this.quitButton_Click);
            // 
            // notifyUserButton
            // 
            this.notifyUserButton.FlatStyle = System.Windows.Forms.FlatStyle.System;
            this.notifyUserButton.Location = new System.Drawing.Point(19, 159);
            this.notifyUserButton.Name = "notifyUserButton";
            this.notifyUserButton.Size = new System.Drawing.Size(154, 47);
            this.notifyUserButton.TabIndex = 2;
            this.notifyUserButton.Text = "&Notify the user when an exception occurs";
            this.notifyUserButton.Click += new System.EventHandler(this.notifyUserButton_Click);
            // 
            // QuickStartForm
            // 
            this.CancelButton = this.quitButton;
            this.ClientSize = new System.Drawing.Size(834, 458);
            this.Controls.Add(this.notifyUserButton);
            this.Controls.Add(this.groupBox);
            this.Controls.Add(this.groupBox1);
            this.Controls.Add(this.logExceptionButton);
            this.Controls.Add(this.resultsTextBox);
            this.FormBorderStyle = System.Windows.Forms.FormBorderStyle.FixedSingle;
            this.MaximizeBox = false;
            this.Name = "QuickStartForm";
            this.StartPosition = System.Windows.Forms.FormStartPosition.CenterScreen;
            this.Text = "Exception Handling Application Block Quick Start";
            this.Load += new System.EventHandler(this.QuickStartForm_Load);
            this.groupBox1.ResumeLayout(false);
            this.groupBox1.PerformLayout();
            ((System.ComponentModel.ISupportInitialize)(this.logoPictureBox)).EndInit();
            this.groupBox.ResumeLayout(false);
            this.ResumeLayout(false);
            this.PerformLayout();

        }
        #endregion

        /// <summary>
        /// The main entry point for the application.
        /// </summary>
        [STAThread]
        static void Main()
        {
            AppForm = new QuickStartForm();
            Application.Run(AppForm);
        }

        private void QuickStartForm_Load(object sender, System.EventArgs e)
        {
            // Initialize image to embedded logo
            logoPictureBox.Image = GetEmbeddedImage("ExceptionHandlingQuickStart.logo.gif");
        }

        private System.Drawing.Image GetEmbeddedImage(string resourceName)
        {
            Stream resourceStream = Assembly.GetEntryAssembly().GetManifestResourceStream(resourceName);

            if (resourceStream == null)
            {
                return null;
            }

            System.Drawing.Image img = System.Drawing.Image.FromStream(resourceStream);

            return img;
        }


        private void DisplayScenarioStart(string scenarioDescription)
        {
            this.resultsTextBox.Text = scenarioDescription + Environment.NewLine + Environment.NewLine;
            this.resultsTextBox.Update();
        }

        private void DisplayResults(string results)
        {
            this.resultsTextBox.Text += results;
        }

        private string GetExceptionInfo(Exception ex)
        {
            StringBuilder sb = new StringBuilder();
            StringWriter writer = new StringWriter(sb);

            AppTextExceptionFormatter formatter = new AppTextExceptionFormatter(writer, ex);

            // Format the exception
            formatter.Format();

            return sb.ToString();
        }

        private void quitButton_Click(object sender, System.EventArgs e)
        {
            this.Close();
        }

        /// <summary>
        /// Routine that causes an exception to be thrown
        /// </summary>
        private void Process()
        {
            throw new Exception("Quick Start Generated Exception");
        }

        private void notifyUserButton_Click(object sender, System.EventArgs e)
        {
            try
            {
                Cursor = System.Windows.Forms.Cursors.WaitCursor;

                StringBuilder sb = new StringBuilder();

                sb.Append("Scenario: Notify the user when an exception occurs");
                sb.Append(Environment.NewLine);
                sb.Append(Environment.NewLine);
                sb.Append("1. An exception occurs and is detected in the Business layer.");
                sb.Append(Environment.NewLine);
                sb.Append(Environment.NewLine);
                sb.Append("2. The Business layer specifies the \"Notify Policy\" as the exception handling policy.");
                sb.Append(Environment.NewLine);
                sb.Append(Environment.NewLine);
                sb.Append("3. The \"Notify Policy\" is configured to first log the exception, then replace the exception with a new one, and finally return to the application by recommending a rethrow.");
                sb.Append(Environment.NewLine);
                sb.Append(Environment.NewLine);
                sb.Append("4. The exception is propagated to and caught by the UI layer.");
                sb.Append(Environment.NewLine);
                sb.Append(Environment.NewLine);
                sb.Append("5. The UI layer catches the exception and calls the \"Global Policy\", which displays the exception in a message box.");
                sb.Append(Environment.NewLine);
                DisplayScenarioStart(sb.ToString());

                AppService svc = new AppService();

                svc.ProcessAndNotify();

                Cursor = System.Windows.Forms.Cursors.Arrow;
            }
            catch (Exception ex)
            {
                ProcessUnhandledException(ex);
            }
        }

        private void logExceptionButton_Click(object sender, System.EventArgs e)
        {
            try
            {
                Cursor = System.Windows.Forms.Cursors.WaitCursor;

                StringBuilder sb = new StringBuilder();

                sb.Append("Scenario: Log exception");
                sb.Append(Environment.NewLine);
                sb.Append(Environment.NewLine);
                sb.Append("1. An exception occurs and is detected in the UI layer.");
                sb.Append(Environment.NewLine);
                sb.Append(Environment.NewLine);
                sb.Append("2. The UI layer specifies the \"Log Only Policy\" as the exception handling policy.");
                sb.Append(Environment.NewLine);
                sb.Append(Environment.NewLine);
                sb.Append("3. The \"Log Only Policy\" is configured to log the exception and return to the application without recommending a rethrow.");
                sb.Append(Environment.NewLine);
                sb.Append(Environment.NewLine);
                sb.Append("4. Control is returned to the UI layer.");
                sb.Append(Environment.NewLine);

                DisplayScenarioStart(sb.ToString());

                try
                {
                    Process();
                }
                catch (Exception ex)
                {
                    bool rethrow = ExceptionPolicy.HandleException(ex, "Log Only Policy");

                    DisplayResults("**Exception has been logged. See the currently configured log destination (default is event log) for exception details.");

                    if (rethrow)
                    {
                        throw;
                    }
                }

                Cursor = System.Windows.Forms.Cursors.Default;
            }
            catch (Exception ex)
            {
                ProcessUnhandledException(ex);
            }

        }

		/// <summary>
		/// Returns the path and executable name for the help viewer.
		/// </summary>
		private string GetHelpViewerExecutable()
		{
            string commonX86 = Environment.GetEnvironmentVariable("CommonProgramFiles(x86)");
            if (!string.IsNullOrEmpty(commonX86))
            {
                string pathX86 = Path.Combine(commonX86, @"Microsoft Shared\Help 9\dexplore.exe");
                if (File.Exists(pathX86))
                {
                    return pathX86;
                }
            }
            string common = Environment.GetEnvironmentVariable("CommonProgramFiles");
            return Path.Combine(common, @"Microsoft Shared\Help 9\dexplore.exe");
        }

        /// <summary>
        /// Displays Quick Start help topics using the Help 2 Viewer.
        /// </summary>
        private void viewWalkthroughButton_Click(object sender, System.EventArgs e)
        {
            // Process has never been started. Initialize and launch the viewer.
            if (this.viewerProcess == null)
            {
                // Initialize the Process information for the help viewer
                this.viewerProcess = new Process();

				this.viewerProcess.StartInfo.FileName = GetHelpViewerExecutable();
                this.viewerProcess.StartInfo.Arguments = HelpViewerArguments;
                this.viewerProcess.Start();
            }
            else if (this.viewerProcess.HasExited)
            {
                // Process previously started, then exited. Start the process again.
                this.viewerProcess.Start();
            }
            else
            {
                // Process was already started - bring it to the foreground
                IntPtr hWnd = this.viewerProcess.MainWindowHandle;
                if (NativeMethods.IsIconic(hWnd))
                {
                    NativeMethods.ShowWindowAsync(hWnd, NativeMethods.SW_RESTORE);
                }
                NativeMethods.SetForegroundWindow(hWnd);
            }
        }

        /// <summary>
        /// Process any unhandled exceptions that occur in the application.
        /// This code is called by all UI entry points in the application (e.g. button click events)
        /// when an unhandled exception occurs.
        /// You could also achieve this by handling the Application.ThreadException event, however
        /// the VS2005 debugger will break before this event is called.
        /// </summary>
        /// <param name="ex">The unhandled exception</param>
        private void ProcessUnhandledException(Exception ex)
        {
            // An unhandled exception occured somewhere in our application. Let
            // the 'Global Policy' handler have a try at handling it.
            try
            {
                bool rethrow = ExceptionPolicy.HandleException(ex, "Global Policy");
                if (rethrow)
                {
                    // Something has gone very wrong - exit the application.
                    Application.Exit();
                }
            }
            catch
            {
                // Something has gone wrong during HandleException (e.g. incorrect configuration of the block).
                // Exit the application
                string errorMsg = "An unexpected exception occured while calling HandleException with policy 'Global Policy'. ";
                errorMsg += "Please check the event log for details about the exception." + Environment.NewLine + Environment.NewLine;

                MessageBox.Show(errorMsg, "Application Error", MessageBoxButtons.OK, MessageBoxIcon.Stop);

                Application.Exit();
            }
            QuickStartForm.AppForm.Cursor = System.Windows.Forms.Cursors.Default;
        }
    }
}
